### ISSUE TITLE
[Youtube] .jpg thumbnails are actually better than .webp

### ISSUE THREAD

**Issue by cpz3501** [yt-dlp/yt-dlp#11242]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a site-specific feature
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Germany

### Example URLs

Any

### Provide a description that is worded well enough to be understood

Maybe I'm missing something ... but for some time now I've noticed that .jpg thumbnails, although having far more artifacts, tend to retain more detail than .webp.

I'm not sure what's causing this, but I'm pretty convinced that .jpg is the superior choice for thumbnails.

Would be nice if someone who is familiar with this kind of stuff could confirm this. To me it looks like youtube did a conversion of `original image -> .jpg -> .webp`.

Just to clarify, .jpg looks worse at first glance, but if you compare closely, you can see that the best .jpg has details that the best .webp thumbnail misses, which is more visible in lower quality uploads.

At the end of the day, this is nitpicking and doesn't really matter, but it would be nice to have the best version possible, so that's why I'm asking this question anyway.

Not a tech guy, this "-vU flag" and verbose stuff is giving me a headache, hopefully you can let it slide? Instead, here is my configuration file, some batch script I copied from stackoverflow. (probably unnecessary for this issue, but maybe someone reading this will find it useful)

```
@echo off
set "url=%~1"
if not defined url set /p "url=Enter URL: "
yt-dlp ^
 --output "D:/xyz/%%(upload_date)s %%(title).100s %%(id)s.%%(ext)s"^
 --download-archive "xyz.txt"^
 --embed-metadata --embed-thumbnail --embed-chapters --embed-subs --sub-langs all,-live_chat^
 --xattrs --merge-output-format mkv^
 %url% 2>>log.txt
timeout 5

```

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
grtrf
```

**Comment by bashonly** [comment]
it would be useful if you could provide any evidence to back up this claim

**Comment by cpz3501** [comment]
@bashonly

Evidence like uploading zoomed in .png screenshots from covers for A/B comparisons? 

Because that's about the most I can do. If you're talking about proper analytics then I can't do it.

**Comment by cpz3501** [comment]
@bashonly
(or anyone who cares enough about a issue that can barely be called one)

So, here we go,

1. Example (official LG 4k HDR demo, very high quality)
Source: https://www.youtube.com/watch?v=njX2bu-_Vw4
JPG: https://i.ytimg.com/vi/njX2bu-_Vw4/maxresdefault.jpg
Webp: https://i.ytimg.com/vi_webp/njX2bu-_Vw4/maxresdefault.webp
Zoom all the way to the center, into the bubble between the lower part of the L and the G letters.
Approximate position: 641, 359px
In .jpg you can see the bubble, whereas in .webp you wouldn't see it if you didn't know.

2. Example (small music channel, medium quality)
Source: https://www.youtube.com/watch?app=desktop&v=ctca-LXU2BQ
JPG: https://i.ytimg.com/vi/ctca-LXU2BQ/maxresdefault.jpg
Webp: https://i.ytimg.com/vi_webp/ctca-LXU2BQ/maxresdefault.webp
Zoom closely into the top right corner, the knight's visor, in the bottom row at the end of the left side of the visor.
Approximate position: 992, 105px
In .jpg you can see that the visor has 2 metal bars at the bottom left, while in webp. it appears as a single blurred bar.

3. Example (very old video, extremely bad quality)
Source: https://www.youtube.com/watch?v=qbdSX8LQDMM
JPG: https://i.ytimg.com/vi_webp/qbdSX8LQDMM/hqdefault.webp
Webp: https://i.ytimg.com/vi_webp/qbdSX8LQDMM/hqdefault.webp
You can see it everywhere, but it's most noticeable on the man in the middle holding the phone.
Approximate position: 267, 147px
In .jpg you can make out that he's holding a phone, while in .webp it almost looks like he's holding a sausage.

I tried to give as detailed and diverse examples as possible. The point I was trying to make is that I believe youtube does something like `original image -> .jpg -> .webp`, meaning that jpg is the better choice at the moment.

My opinion on this is, although JPG has far more visible artifacts, I think most people prefer the higher detail over the low quality "clean up" that happens as a side effect of webp compression.
What's also worth noting is the artifacts are only visible when zoomed in, during normal viewing JPG thumbnails are the better viewing experience, especially on high-resolution displays with decent built-in upscaling, such as the newer LG 65" TVs.

As far as my test method is concerned:

Visible Comparisson:
Firefox (standard image view, cross-checking)
Nomac (image viewer, zoomed A/B change, cross comparison)
Imageglass (image viewer, zoomed A/B change)

Pixel position estimation:
Windows Paint (the number in the lower left corner)


Sidenote:
English is not my native language and I used translation software for assistance.

**Comment by bashonly** [comment]
I can definitely see a difference in examples 2 and 3 (jpg looks better than webp), but not so much in example 1.

I buy the "jpg is better quality" argument, and if quality was the only consideration, I wouldn't hesitate to change the default to jpg.

But what gives me pause is the technical repercussions of doing so. This would negatively impact users whose configs/scripts depend on the thumbnail being webp, e.g. if they are using `--convert-thumb jpg` and `--ppa ThumbnailsConvertor:...` (which is [not uncommon](https://github.com/yt-dlp/yt-dlp/issues/429#issuecomment-2282554440)), then `--convert-thumb` becomes a no-op and their config breaks.

I'd like to keep this open for discussion for a while, and I've added this to the [DISCUSSIONS post](https://github.com/yt-dlp/yt-dlp/issues/3765) to bring some more visibility to this issue and hopefully get some more voices to chime in

**Comment by fireattack** [comment]
I agree with OP; in general YouTube's jpeg thumbnail is better than the webp's counterpart because they use very aggressive compression ratio for their webp encoder (which unfortunately is a very common practice in various image hosting services, and IMHO defeats the points of using webp partially).

I personally don't even need yt-dlp to switch jpeg by default; but there doesn't seem to be able to manually set it to download jpeg instead, at all, last time I asked in Discord.

**Comment by bashonly** [comment]
> I personally don't even need yt-dlp to switch jpeg by default; but there doesn't seem to be able to manually set it to download jpeg instead, at all, last time I asked in Discord.

Correct, at least not until #237 is resolved. IIRC pukkandan said *somewhere* (I can't find it now) that when [format selection is reworked](https://github.com/yt-dlp/yt-dlp/issues/4846), thumbnail selection would become easier to implement. Not sure when that will happen though

In the meantime, maybe we could add an extractor-arg to control this? e.g. `--extractor-arg "youtube:preferred_thumbnail_ext=jpg"` or something like that. Then we wouldn't even need to change the default

**Comment by francoism90** [comment]
I personally found myself webp isn't that great when it comes to quality of an image. It's compression is also pretty sensitive, you could end up with a larger file compared to jpg. The only benefit seems to be that webp could be faster generated/rendered for a browser.

I think that's why Google is looking at jpegxl again, and other formats that are new like AFIV.

**Comment by ggtylerr** [comment]
I'm not sure about qualities between formats, but if it's of any help, KDE's Dolphin / ffmpegthumbs doesn't support webp covers. I don't know of any other examples, but I think for compatibility's sake, jpg / png should be used.

**Comment by lisansas** [comment]
I think jpg is superior to webp both in terms of compatibility and image quality

**Comment by 0n-s** [comment]
A coincidental effect of this would be, when muxing to MKV files, being compliant with [the standard regarding thumbnails](https://www.matroska.org/technical/attachments.html) (I have no idea if this extends to WebM). This is important since FFmpeg & other software (e.g. GStreamer) only consider an image attachment to be a thumbnail if it matches the standard (i.e. PNG or JPEG). This means videos downloaded using yt-dlp can get the real thumbnails in file managers again.

I mention the standard since it's not just a matter of "legacy compatibility"; anything other than those formats is a nonstandard quirk you cannot rely on, regardless of whether the OS at large supports any of those formats.

---

Also, +1 on this idea. Modern codecs (WebP/HEVC/etc.) tend to blur a lot at low bitrates where older codecs (H.264/JPEG/etc.) would artifact a lot, but actually retain details better, & YouTube thumbnails are a great example of that (both in the examples cited & my own experience). Which is unfortunate since WebP is much nicer at higher bitrates, but YouTube evidently heavily skimps on quality even for thumbnails.

**Comment by tcely** [comment]
I also prefer `.jpg` thumbnails from YouTube.

Switching the default and ensuring that `yt_dlp` doesn't break when a post processing step becomes unnecessary would be best, in my opinion.

If that won't happen immediately, or at all, then using the extractor argument is better than the current situation.

> In the meantime, maybe we could add an extractor-arg to control this? e.g. --extractor-arg "youtube:preferred_thumbnail_ext=jpg" or something like that.

**Comment by pukkandan** [comment]
> Correct, at least not until #237 is resolved. IIRC pukkandan said _somewhere_ (I can't find it now) that when [format selection is reworked](https://github.com/yt-dlp/yt-dlp/issues/4846), thumbnail selection would become easier to implement. Not sure when that will happen though

For a full-fledged selector (with filters etc), the format rework is needed. But it would be possible to implement a minimal selector like `--thumb-format jpg` for the time being and extend it later

**Comment by uaevuon** [comment]
omg, it really is. jpg thumbnails are superior in terms of quality.

**Comment by LaloSquid** [comment]
I also did some testing and can confirm that the jpg thumbnails have superior quality compared to the webp variants.
For example have a look at these videos, https://youtu.be/tMn5vpZofUE and https://youtu.be/Rkbb-llcLrQ using `--list-thumbnails`:

https://slow.pics/c/lKJZTO6z (Image comparison site)
For the thumbnail with the car, look in the reflection detail near the rear tire, and overall detail on the front tire.
For the thumbnail about movies, look at the grain retention.

Despite the maxresdefault.webp variant showing up as "1080p" in the list when doing `--list-thumbnails`, it is 720p just like the jpg variant. I suspect that this is the same across all videos but I have only done limited testing, so I don't know.

I would like to propose that the default is changed, so that the jpg thumbnails are downloaded by default when using `--embed-thumbnail` from YouTube. Currently I see no way of choosing what thumbnail format to download in (If such an option exists, please let me know!), but only an option to convert thumbnails after they have been grabbed. But this conversion is unfortunately implemented in a lossy way, degrading quality significantly further.

As a motivation, I would like to note that I have been facing difficulties getting webp thumbnails to show up in various file explorers on linux, but jpg works without issue.

**Comment by atsushi2965** [comment]
> > I personally don't even need yt-dlp to switch jpeg by default; but there doesn't seem to be able to manually set it to download jpeg instead, at all, last time I asked in Discord.
> 
> Correct, at least not until [#237](https://github.com/yt-dlp/yt-dlp/issues/237) is resolved. IIRC pukkandan said _somewhere_ (I can't find it now) that when [format selection is reworked](https://github.com/yt-dlp/yt-dlp/issues/4846), thumbnail selection would become easier to implement. Not sure when that will happen though
> 
> In the meantime, maybe we could add an extractor-arg to control this? e.g. or something like that. Then we wouldn't even need to change the default`--extractor-arg "youtube:preferred_thumbnail_ext=jpg"`

how about forcing it to download jpg and adding compat-option for webp instead

**Comment by bashonly** [comment]
> how about forcing it to download jpg and adding compat-option for webp instead

There have been a couple of site-specific compat options added in the past (`no-youtube-channel-redirect` and `no-youtube-unavailable-videos`), but IMO doing so was a mistake and I would strongly prefer to avoid adding any more site-specific compat opts. I think compat opts should be reserved for *general* backwards compatibility only.

For example: What if we change the Youtube extractor to default to `jpg` thumbnails and add a `youtube-webp-thumbnails` compat option, but then Youtube migrates fully to `jpg` (or some other format) and removes all of the `webp` thumbnails? We'll have an obsolete compat opt that we can't remove (since compat opts are forever). And it will be worse than obsolete: it will mislead the user into thinking it can do something that it can't.

In general, yt-dlp does not promise compatibility at the extractor level because most extractor level functionality is ultimately out of our control. We are at the whim of the site. 

At the same time, the Youtube extractor is one of the most commonly used extractors in the project, and as a maintainer I would like to avoid causing any drastic differences between the user's expected results and the actual results, since there is already enough of that caused by the sites.

Ideally, a general option for thumbnail selection would solve this problem. That could take a while though, since we'd need to arrive at a consensus on the design for thumbnail selection, and, depending on what the design is, it could take some effort to implement.

In the meantime, I would be OK with adding a Youtube `--extractor-arg` to configure what thumbnail format is preferred, since we do not guarantee any backwards compat for `--extractor-args` and can add/remove/change them at will.

**Comment by tcely** [comment]
Any improvement regarding how thumbnails are being handled would be welcomed by me.

I'd prefer to see the default for YouTube become `.jpg`. But, however many people agree with me, I still want those who prefer `.webp` to have the ability to download too.

I think an extractor argument can easily allow both. To me that's good enough for now, until a more general option becomes available.

**Comment by cynthia2006** [comment]
In a summary the issue here is the following.

WebP is based on the lossy VP8 video codec, which is old according to today's standards. In 2010, Fiona Glasser (a x264 developer) wrote a [critique of WebP](https://web.archive.org/web/20150319214453/http://x264dev.multimedia.cx/archives/541). The use of video codecs as image formats is suboptimal when it comes to detail retention. Video codecs tend to blur details based on how our visual stimulus functions; a frame barely lasts for 30ms (for a 30 FPS) video, so the blurring is unnoticeable. However, modern codecs like AV1 support Film Grain Synthesis (FGS), which analyses (and optionally denoises) film grain, and synthesizes it at the decoding site. Formats like AVIF can technically leverage this benefit to prevent over-smoothing of detail. However, Google doesn't support AVIF — only WebP and JPEG. And, Google although being one of the major contributors to JPEG-XL — a format meant to be a successor of JPEG, and obviously better than WebP — shows a reluctance to include it in their ecosystem, partly to push the WebP agenda forward.

However, my opinion of this issue is the following. A minority of people care about embeddeding thumbnails for audio/videos. More so, thumbnails in the file manager or players will almost be shown at a much smaller size, so the difference in detail retention won't be noticeable. Therefore, it's best to leave the behaviour unchanged, and instead focus on stability instead.

**Comment by nicolaasjan** [comment]
> And, Google although being one of the major contributors to JPEG-XL — a format meant to be a successor of JPEG, and obviously better than WebP — shows a reluctance to include it in their ecosystem, partly to push the WebP agenda forward.

(offtopic)
I have good news for you. 😀️

[JPEG-XL Image Support Returns To Latest Chrome / Chromium Code](https://www.phoronix.com/news/JPEG-XL-Returns-Chrome-Chromium)

**Comment by uaevuon** [comment]
> However, my opinion of this issue is the following. A minority of people care about embeddeding thumbnails for audio/videos. More so, thumbnails in the file manager or players will almost be shown at a much smaller size, so the difference in detail retention won't be noticeable. Therefore, it's best to leave the behaviour unchanged, and instead focus on stability instead.

This actually affects --write-thumbnail as well, which has nothing to do with embedding thumbnails.

**Comment by eladkarako** [comment]
note:

in the current command line:

`ffmpeg -y -loglevel repeat+info -f image2 -pattern_type none -i 'file:input.webp' -update 1 -bsf:v mjpeg2jpeg -movflags +faststart 'file:output.jpg'`

in webp, png, svg - transparent pixels (alpha channel) will be converted to black pixels.

this will place a white background behind the image:

`ffmpeg -y -loglevel "repeat+info" -f lavfi -i "color=#FFFFFF" -f image2 -pattern_type none -i "file:input.webp" -update 1 -bsf:v mjpeg2jpeg -filter_complex "[0][1]scale2ref[bg][fg];[bg][fg]overlay=shortest=1,format=yuvj420p" -movflags "+faststart" "file:output.jpg"`

note this not really an issue for youtube,  
or if the thumbnail is a frame from an actual video (most cases probably).  

I just want to put this in here  
just in case anyone is seeing weird black pixels in their thumbnail ;)  

optionally adding `"-frames:v" 1` to restrict to one frame encoding,  
and `-map_metadata "-1"` to prevent copying meta-data from the original thumbnail to the jpg, since it is attached to container, and can be easily extracted afterwards.

**Comment by vt-idiot** [comment]
Why is there no way to select the thumbnail format in the first place? Why is it defaulting to WEBP? YouTube's WEBP thumbnails are awful

**Comment by cynthia2006** [comment]
Honestly, why are people so worrying about thumbnails anyway? It's not as if you're going to stare at it more than a second. Besides, yes, WebP is known to be blurry because it uses VP8 under the hood, which is a video codec and wasn't designed with images in mind. Fiona Glaser wrote a critique of it back in 2010. Her article [here](https://web.archive.org/web/20150319214453/http://x264dev.multimedia.cx/archives/541).

**Comment by 0n-s** [comment]
> Honestly, why are people so worrying about thumbnails anyway? It's not as if you're going to stare at it more than a second.

Given the dates between messages I find it hard to say people are "so worrying", but here are some more important arguments:

When looking at a directory of videos in a file manager, it means your file manager gets to just snag an image out of the container and resize it, rather than run ffmpeg on all cores decoding god knows how many frames to find a representative sample to generate an image that the video itself already provided.

This really does just come down to how WebP is not a standardized format for thumbnails in Matroska (and by extension WebM in practice), as elucidated in more detail [here](https://github.com/yt-dlp/yt-dlp/issues/11242#issuecomment-2521306512).

> Besides, yes, WebP is known to be blurry

This sounds like a great argument for this issue to be resolved. YouTube provides JPEG thumbnails for every video, and yt-dlp has graceful fallback in case they ever don't. I've been running a patch that inverts the priority of JPEG and WebP selection for YouTube for the past couple of years or so with no problem, with lots of videos downloaded per day. It works consistently.

And among yt-dlp users, there are a disproportionate amount of archivists which probably would like for the program to select the highest quality thumbnail format available automatically, and that's very rarely the WebP ones.

**Comment by cynthia2006** [comment]
> ...your file manager gets to just snag an image out of the container and resize it, rather than run ffmpeg on all cores decoding god knows how many frames to find a representative sample...

Although it's certainly true that without providing a thumbnail in the container files require extra processing, a file-manager (specifically `ffmpegthumbs`) doesn't have to decode the entire video, nor read the entire file. `libavformat` performs **container-level seeks** by default, which is incredibly cheap, so the file manager *skims* over the contents of the file—decoding one or more intra-frames—which is possible without decoding any other frames in the GOP, and so it's cheap as well. Furthermore, the file manager almost certainly has a cache where it stores the thumbnails so as to not repeat the task. Yes, it's more expensive but your argument is quite hyperbolic.

A better argument is that, thumbnails are better representative of the videos than whatever the file manager generates.

Personally, I'm averse to WebP images because, although they're substantially smaller in size, they're quite blurry because of the underlying mechanics of the format. And although I'm a contender of JXL, Google hasn't implemented it into their infrastructure yet (which is ironic because the ISO standardised format is based on Google's PIK and Cloudinary's FUIF), and thus my choice is JPEG as well; it's universally supported plus it's sharp enough.

**Comment by 0n-s** [comment]
> Yes, it's more expensive but your argument is quite hyperbolic.

Yes, those points are correct. To that, I will just say that tone doesn't carry over well in pure text. Nevertheless I think the point I made is a good enough argument for an issue that can be solved like this:
```
diff --git i/yt_dlp/extractor/youtube/_video.py w/yt_dlp/extractor/youtube/_video.py
index 8c5fa5593..6c30c7458 100644
--- i/yt_dlp/extractor/youtube/_video.py
+++ w/yt_dlp/extractor/youtube/_video.py
@@ -4113,7 +4113,7 @@ def feed_entry(name):
         } for name in thumbnail_names for ext in ('webp', 'jpg'))
         for thumb in thumbnails:
             i = next((i for i, t in enumerate(thumbnail_names) if f'/{video_id}/{t}' in thumb['url']), n_thumbnail_names)
-            thumb['preference'] = (0 if '.webp' in thumb['url'] else -1) - (2 * i)
+            thumb['preference'] = (-1 if '.webp' in thumb['url'] else 0) - (2 * i)
         self._remove_duplicate_formats(thumbnails)
         self._downloader._sort_thumbnails(original_thumbnails)
 
```
This would result in better behaviour immediately even without implementing any user-facing format selection, as I've yet to encounter a case where the WebP thumbnail is the higher quality one. Thumbnail format selection is fine and all, but that's a broader architectural point, whereas this issue is more about the sort order for thumbnails being wrong. More about good defaults, not choices.

**Comment by cynthia2006** [comment]
Indeed, JPEG should be the preferred default and WebP should be available through thumbnail format selection.

**Comment by tcely** [comment]
> Indeed, JPEG should be the preferred default and WebP should be available through thumbnail format selection.

Those are the exact goals of: https://github.com/yt-dlp/yt-dlp/pull/17280

If anyone sees anything that must be changed in that pull request, then please comment on it.

**Comment by cdwm** [comment]
Some measured bytes for this thread and for #17280, since most of the comparisons above are single videos. On 19 September 2026 I sent HEAD requests for `maxresdefault.webp`, `sddefault.webp` and `hqdefault.webp` (under `i.ytimg.com/vi_webp/`) for 1,000 public videos that have a `maxresdefault.jpg` (500 uploads from 2018 to 2026, 500 older Wikidata-referenced videos) and paired them with the JPEG sizes of the same files.

| size | WebP exists | median JPEG | median WebP | median WebP/JPEG | WebP larger than JPEG |
| --- | --- | --- | --- | --- | --- |
| maxresdefault 1280x720 | 965 | 96.7 KB | 50.8 KB | 0.53 | 1.0% |
| sddefault 640x480 | 970 | 41.3 KB | 23.3 KB | 0.55 | 1.4% |
| hqdefault 480x360 | 970 | 19.0 KB | 15.6 KB | 0.75 | 33.4% |

The hqdefault row hides an age split. For the older videos the median ratio is 0.65 and WebP is larger for 10%. For the 2018 to 2026 uploads it is 1.06 and WebP is larger for 56%, rising by upload year: 25% of 2020 uploads, 43% of 2024, 55% of 2025, 72% of 2026 (n = 217, median ratio 1.13). Vertical Shorts are the extreme at hqdefault (WebP larger for 85%, median 1.17) while their maxresdefault WebP is 0.40 of the JPEG.

Two caveats. Bytes are not quality: these are separate encodes of the same frame and I have no fidelity measure, so a heavier WebP may just be a better encode; the table only settles which file is lighter. And a missing WebP is HTTP 404 with a 552-byte `image/webp` body (30 of the 1,000 had no WebP at any size, 5 more had sd/hq but no maxres), the same shape as the 120x90 JPEG placeholder, so any preference logic needs the status check before the format check.

Method, limitations and the per-video CSV (CC BY 4.0): https://thumbnailsgrabber.com/youtube-thumbnail-statistics#webp-2026-09

**Comment by tcely** [comment]
Interesting.

One thing bothered me when reading; a lot of knowledge was assumed when discussing comparisons.

For example: listing the actual byte size of the placeholder images in the 404 response, when discussing them, would be very helpful.

Also, a table mapping resolutions to the various filenames would be a handy reference.

### HTTP 404 Placeholder

Format | Size | Resolution
-------|------|-----------
`JPEG` [^jpg] | `1,097` | `120x90`
`WebP` [^webp] | `552` | `120x90`

[^jpg]: placeholder.jpg: JPEG image data, JFIF standard 1.01, aspect ratio, density 1x1, segment length 16, baseline, precision 8, 120x90, components 3

[^webp]: placeholder.webp: RIFF (little-endian) data, Web/P image, VP8 encoding, 120x90, Scaling: [none]x[none], YUV color, decoders should clamp


### Resolution and requested file name

Resolution | `maxresdefault` | `sddefault` | `hqdefault`
----------- | -------------- | ----------- | ---------
`1920x1080` [^1080p] | ✅ | - | -
`1280x720` | ✅ | - | -
`640x480` | - | ✅ | -
`480x360` | - | - | ✅

[^1080p]: Only observed on older videos.

**Comment by cdwm** [comment]
You beat me to the placeholder sizes. A few extra rows from the same probe, since the grid only covers the middle three files and says nothing about how often each one is actually there (missing = 404 with that placeholder):

| File | Pixels | Exists (8,664 videos) |
| --- | --- | --- |
| default.jpg | 120x90 | 100% |
| mqdefault.jpg | 320x180 | 100% |
| hqdefault.jpg | 480x360 | 100% |
| sddefault.jpg | 640x480 | 95.8% |
| hq720.jpg | 1280x720 | 88.2% |
| maxresdefault.jpg | 1280x720 | 88.3% |
| vi_webp/.../maxresdefault.webp | 1280x720 | 85.7% |
| oar2.jpg | original aspect, usually 1080x1920 on Shorts | 29.9% |

Three notes on the table.

The 1080p footnote is right. 69 videos in the sample returned a real 1920x1080 from the maxresdefault URL. All of them were older uploads. Zero of the 4,607 recent ones did.

hq720.jpg is basically an alias for maxresdefault.jpg. It never existed for a video that lacked maxres, and 93.6% of the time it was byte for byte the same file.

The sd and hq WebP files live under the same vi_webp/ path, same filenames as the JPEGs.

If there's anything else worth checking against the same set for the PR, say so and I'll run it.

**Comment by tcely** [comment]
Thanks.

The approach that I was using mirrors the recommendation:
1. grab the hq default
2. try upgrading by fetching sd default
3. try upgrading by fetching maxres default

This procedure could be optimized a bit more, but simply scheduling those tasks has been good enough to consistently retrieve and store a thumbnail, at the resolution defined by the user. The task will scale down to the requested resolution for storage when the download is larger.

**Comment by cynthia2006** [comment]
@cdwm Perhaps it would have been more helpful if you used a perceptual quality metric such as [SSIMULACRA2](https://github.com/cloudinary/ssimulacra2). Assuming JPEG as the superior, take it as a reference then measure the distortion of the WebP. This would shed some light on the codec efficiency.
